models/protocol.py

```diff
@@ -66,6 +66,26 @@ def remove(session, bytes_identifier, timestamp):
 
         protocol.coverage_ended_at = datetime.fromtimestamp(timestamp)
 
+    @staticmethod
+    def get_sum_of_premiums(session):
+        # Sums up the latest premium of each protocol.
+        sql = text(
+            """
+            SELECT SUM(pc.premium) premiums
+            FROM protocols_premiums pc
+            INNER JOIN (
+                SELECT p.id, MAX(pc.premium_set_at) timestamp
+                FROM protocols p
+                JOIN protocols_premiums pc ON pc.protocol_id = p.id
+                WHERE p.coverage_ended_at IS NULL
+                GROUP BY p.id
+            ) max ON max.id = pc.protocol_id AND max.timestamp = pc.premium_set_at"""
+        )
+
+        result = session.execute(sql).first()
+
+        return result["premiums"]
+
```

Comment:
coverage_ended_at can only be null on a `ProtocolRemoved` or `ProtocolRemovedByArb` event. In which case the premium will also be set to zero. It's sufficient to look at the last premium row.

This is the simples query I could come up with

```
SELECT sum(p.premium) FROM ( 
	SELECT 
	DISTINCT ON (protocol_id) premium 
	FROM protocols_premiums 
	ORDER BY protocol_id, premium_set_at 
) as p;
```

Reply:
The order must be `protocol_id, premium_set_at DESC` in order to pick the latest premium.

Can be written using the ORM as well, instead of raw SQL:

```python
session.query(sa.func.sum(ProtocolPremium.premium)).filter(
    ProtocolPremium.id.in_(
        session.query(ProtocolPremium.id)
        .distinct(ProtocolPremium.protocol_id)
        .order_by(ProtocolPremium.protocol_id, ProtocolPremium.premium_set_at.desc())
    )
).scalar()

```

Reply:
Nice, I like the ORM solution

Reply:
`DISTINCT` != `DISTINCT ON` https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-select-distinct/